### PR TITLE
Add HermitianPositiveSemidefiniteConeTriangle with bridges

### PR DESCRIPTION
**Pull request by blegat**
Refactored from https://github.com/jump-dev/ComplexOptInterface.jl

### CODE REVIEW

File: src/Bridges/Variable/bridges/hermitian.jl [hunk 1]
```diff
@@ -0,0 +1,350 @@
+# Copyright (c) 2017: Miles Lubin and contributors
+# Copyright (c) 2017: Google Inc.
+#
+# Use of this source code is governed by an MIT-style license that can be found
+# in the LICENSE.md file or at https://opensource.org/licenses/MIT.
+
+const EQ{T} = MOI.ConstraintIndex{MOI.ScalarAffineFunction{T},MOI.EqualTo{T}}
```
Comment (odow) [inline src/Bridges/Variable/bridges/hermitian.jl:7]: Please don't re-introduce things like this. It just makes everything harder to read. Verbosity is okay.

File: src/Bridges/Variable/bridges/hermitian.jl [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+    return HermitianToSymmetricPSDBridge(
+        variables,
+        psd_constraint,
+        con_11_22,
+        con12diag,
+        con_12_21,
+    )
+end
+
+const HermitianToSymmetricPSD{T,OT<:MOI.ModelLike} =
```
Comment (odow) [inline src/Bridges/Variable/bridges/hermitian.jl:156]: This is duplicated

File: src/Bridges/Variable/bridges/hermitian.jl [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+    SingleBridgeOptimizer{HermitianToSymmetricPSDBridge{T},OT}
+
+function supports_constrained_variable(
+    ::Type{<:HermitianToSymmetricPSDBridge},
+    ::Type{MOI.HermitianPositiveSemidefiniteConeTriangle},
+)
+    return true
+end
+function MOI.Bridges.added_constrained_variable_types(
```
Comment (odow) [inline src/Bridges/Variable/bridges/hermitian.jl:165]:
```suggestion

function MOI.Bridges.added_constrained_variable_types(
```
Pet peeve: spaces between function definitions.

File: src/Bridges/Variable/bridges/hermitian.jl [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+)
+    return bridge.variables
+end
+function MOI.get(
+    bridge::HermitianToSymmetricPSDBridge,
+    ::MOI.NumberOfConstraints{
+        MOI.VectorOfVariables,
+        MOI.PositiveSemidefiniteConeTriangle,
+    },
+)
```
Comment (odow) [inline src/Bridges/Variable/bridges/hermitian.jl:192]:
```suggestion
)::Int64
```

File: src/Bridges/Variable/bridges/hermitian.jl [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+    for ci in bridge.con_11_22
+        MOI.delete(model, ci)
+    end
+    for ci in bridge.con12diag
+        MOI.delete(model, ci)
+    end
+    for ci in bridge.con_12_21
+        MOI.delete(model, ci)
+    end
+    return MOI.delete(model, bridge.variables)
```
Comment (odow) [inline src/Bridges/Variable/bridges/hermitian.jl:230]:
```suggestion
    MOI.delete(model, bridge.variables)
    return
```

File: src/Bridges/Variable/bridges/hermitian.jl [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+    return mapped
+end
+
+function MOI.get(
+    model::MOI.ModelLike,
+    attr::MOI.VariablePrimal,
+    bridge::HermitianToSymmetricPSDBridge{T},
+    i::MOI.Bridges.IndexInVector,
+) where {T}
+    return value = MOI.get(model, attr, _variable(bridge, i))
```
Comment (odow) [inline src/Bridges/Variable/bridges/hermitian.jl:333]:
```suggestion
    return MOI.get(model, attr, _variable(bridge, i))
```

File: src/Test/test_conic.jl [hunk 7]
```diff
@@ -6831,3 +6831,165 @@ function setup_test(
...
+```
+[(1+√3)/2  -1/2+im/2]
+[-1/2+im/2 (-1+√3)/2]
+```
+"""
+function test_conic_HermitianPositiveSemidefiniteConeTriangle_1(optimizer::MOI.ModelLike, config::Config{T}) where {T}
+    atol = config.atol
+    rtol = config.rtol
+
+    MOI.empty!(optimizer)
```
Comment (odow) [inline src/Test/test_conic.jl:6865]:
```suggestion
```

You don't need this

File: src/Test/test_conic.jl [hunk 8]
```diff
@@ -6831,3 +6831,165 @@ function setup_test(
...
+            primal[3] + one(T),
+            √T(2) * (primal[4] - one(T)),
+        ]
+        t_value = LinearAlgebra.norm(soc_primal)
+        soc_primal = [t_value; soc_primal]
+        dual = [(T(3) - √T(3)) / T(6), √T(3)/T(6), (T(3) + √T(3)) / T(6), -√T(3)/T(6)]
+        @test MOI.Utilities.set_dot(primal, dual, set) ≈ 0 atol = atol rtol = rtol
+        soc_dual = [one(T), -dual[1], -dual[2]*√T(2), -dual[3], -dual[4]*√T(2)]
+        @test MOI.Utilities.set_dot(soc_primal, soc_dual, soc_set) ≈ 0 atol = atol rtol = rtol
+        @test MOI.get(optimizer, MOI.VariablePrimal(), x) ≈ primal atol = atol rtol =
```
Comment (odow) [inline src/Test/test_conic.jl:6902]:
```suggestion
        @test ≈(MOI.get(optimizer, MOI.VariablePrimal(), x), primal, config)
```

I added a method to `config` to simplify these sorts of lines.

File: src/Test/test_conic.jl [hunk 9]
```diff
@@ -6831,3 +6831,165 @@ function setup_test(
...
+            rtol
+        @test MOI.get(optimizer, MOI.ConstraintPrimal(), con_soc) ≈ soc_primal atol = atol rtol =
+            rtol
+        if _supports(config, MOI.ConstraintDual)
+            @test MOI.get(optimizer, MOI.ConstraintDual(), cx) ≈ dual atol = atol rtol =
+                rtol
+            @test MOI.get(optimizer, MOI.ConstraintDual(), con_soc) ≈ soc_dual atol = atol rtol =
+                rtol
+        end
+    end
```
Comment (odow) [inline src/Test/test_conic.jl:6915]:
```suggestion
    end
    return
```

File: src/sets.jl [hunk 10]
```diff
@@ -795,6 +795,18 @@ function triangular_form(::Type{PositiveSemidefiniteConeSquare})
     return PositiveSemidefiniteConeTriangle
 end
 
+struct HermitianPositiveSemidefiniteConeTriangle <: AbstractVectorSet
```
Comment (odow) [inline src/sets.jl:798]: Needs docstring and to be added to docs

**Review by odow — COMMENTED** [review]
I didn't dig into the reformulation. But I don't think we want to rush this into MOI v1.7. It's quite complicated.

**Comment by blegat** [comment]
Ok, I'll merge https://github.com/jump-dev/MathOptInterface.jl/pull/1959 then

File: test/Bridges/Variable/hermitian.jl [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        MOI.Test.Config(),
+    )
+    return
+end
+
+function test_runtests()
+    MOI.Bridges.runtests(
+        MOI.Bridges.Variable.HermitianToSymmetricPSDBridge,
+        """
+        constrainedvariable: [r11, r12, r22, c12] in HermitianPositiveSemidefiniteConeTriangle(2)
```
Comment (blegat) [inline test/Bridges/Variable/hermitian.jl:89]: Thanks! Any objection to merge?

**Comment by chriscoey** [comment]
This PR should allow me to add direct support for this MOI Hermitian PSD cone to Hypatia's MOI interface

**Review by odow — APPROVED** [review]

**Comment by odow** [comment]
Actually, just hold off. I'm tidying the docstrings.

**Comment by odow** [comment]
@blegat take another look. I fleshed out the docstring with an example of the reformulation. I think it makes things a bit clearer. 

I also removed your note on duality. It was a bit confusing and out of place. Once you see the example of the reformulation it becomes obvious why you can just sum up the elements. The equality constraints are just enforcing parts of the matrix, not anything structural.

File: src/Bridges/Variable/bridges/hermitian.jl [hunk 12]
```diff
@@ -27,63 +27,48 @@
...
-Moreover, as `(S ∩ T)* = S* + T*` for cones `S` and `T`, we have
-```
-P* = S* + con_11_22* + con_12_21* + con12diag*
-```
-the dual vector of `P*` is the dual vector of `S*` for which we add in the
-corresponding entries the dual of the three constraints, multiplied by the
-coefficients for the `EqualTo` constraints.
+is positive semidefinite.
 
-Note that these contributions cancel out when we multiply them by `A*`:
```
Comment (blegat) [inline src/Bridges/Variable/bridges/hermitian.jl]: I'd prefer keeping these duality notes. If you don't want to put them in the docstring, we can just add them as comment above `ConstraintDual`

**Comment by blegat** [comment]
> Once you see the example of the reformulation it becomes obvious why you can just sum up the elements. The equality constraints are just enforcing parts of the matrix, not anything structural.

Why is that obvious that you don't have to include the dual of `ceq` ? The way I convince myself is as follows:
For the entry `x_{11}`, we have `P_{11} - P_{'44}` as constraints so if `\lambda` is the dual of this constraint the dual at `11` gets `\lambda` and at `44`, it gets `-\lambda`. But the dual obtained by multiplying by `A*` so it's the sum of the dual at `11` + the dual at `44` so you get `\lambda - \lambda = 0`.
Is there an easier way to get this ?

**Comment by odow** [comment]
I didn't try to prove haha. It just seemed right. We don't actually implement the reformulated PSD matrix, so in some sense, the equality constraints are a lower-level implementation detail. What we actually ask for is the duals on that PSD matrix.

File: src/Bridges/Variable/bridges/hermitian.jl (hunk 12 above)
Comment (odow) [reply src/Bridges/Variable/bridges/hermitian.jl]: Yeah, they shouldn't go in the docstring. It's not really meant for consumers of the bridge, just for people wanting to grok the particular implementation of the function.

File: src/Test/test_conic.jl [hunk 13]
```diff
@@ -6831,3 +6831,214 @@ function setup_test(
...
+(1 + √3) / 4   * [1-√3]
+```
+which is
+```
+[(1+√3)/2  -1/2+im/2]
+[-1/2+im/2 (-1+√3)/2]
+```
+"""
+function test_conic_HermitianPositiveSemidefiniteConeTriangle_1(
+    optimizer::MOI.ModelLike,
```
Comment (odow) [inline src/Test/test_conic.jl:6865]: You've mixed `model` and `optimizer`

File: src/Test/test_conic.jl [hunk 14]
```diff
@@ -6831,3 +6831,206 @@ function setup_test(
...
+        ),
+    )
+    model.eval_variable_constraint_dual = false
+    return () -> model.eval_variable_constraint_dual = true
+end
+
+function version_added(
+    ::typeof(test_conic_HermitianPositiveSemidefiniteConeTriangle_2),
+)
+    return sv"1.7.0"
```
Comment (odow) [inline src/Test/test_conic.jl:7035]:
```suggestion
    return v"1.7.0"
```

**Review by odow — APPROVED** [review]

File: src/Test/test_conic.jl [hunk 15]
```diff
@@ -6831,3 +6831,206 @@ function setup_test(
...
+    @requires MOI.supports_constraint(model, MOI.VariableIndex, MOI.EqualTo{T})
+    set = MOI.HermitianPositiveSemidefiniteConeTriangle(3)
+    x, cx = MOI.add_constrained_variables(model, set)
+    primal = T[1, 0, 1, 0, 0, 0, -1, 0, 0]
+    MOI.add_constraints(model, x, MOI.EqualTo.(primal))
+    MOI.optimize!(model)
+    @test ≈(MOI.get(model, MOI.VariablePrimal(), x), primal, config)
+    @test ≈(MOI.get(model, MOI.ConstraintPrimal(), cx), primal, config)
+    if _supports(config, MOI.ConstraintDual)
+        dual = zeros(T, 9)
```
Comment (odow) [inline src/Test/test_conic.jl:7006]: @blegat, some solvers are failing this test: https://github.com/jump-dev/SolverTests/actions/runs/2857418874

Is it possible for the dual to be non-zero?

**Pull request merged**